modcc/modcc.cpp

```diff
@@ -188,99 +184,96 @@ int main(int argc, char **argv) {
-                return report_error("empty file: "+modfile);
-            }
+            if (m.empty()) return report_error(fmt::format("Input file is empty: {}", modfile.string()));
 
             // Perform parsing and semantic analysis passes.
-
             emit_header("parsing");
             Parser p(m, false);
-            if (!p.parse()) {
-                // Parser::parse() writes its own errors to stderr.
-                return 1;
-            }
+            if (!p.parse()) return 1;
 
             emit_header("semantic analysis");
             m.semantic();
             if (m.has_warning()) {
-                cerr << yellow("Warnings:\n");
-                cerr << m.warning_string() << "\n";
-            }
-            if (m.has_error()) {
-                return report_error(m.error_string());
+                std::cerr << yellow("Warnings:\n")
+                          << m.warning_string() << "\n";
             }
 
-            // Generate backend-specific sources for each backend provided.
+            if (m.has_error()) return report_error(m.error_string());
 
+            // Generate backend-specific sources for each backend provided.
             emit_header("code generation");
 
-            std::string prefix = m.module_name();
-            if (!opt.outprefix.empty()) {
-                if (opt.outprefix.back() != '/') opt.outprefix += "/";
-                prefix = opt.outprefix + prefix;
-            }
+            std::string mod = m.module_name();
 
             bool have_cpu = opt.targets.find(targetKind::cpu) != opt.targets.end();
             bool have_gpu = opt.targets.find(targetKind::gpu) != opt.targets.end();
 
-            io::write_all(build_info_header(m, popt, have_cpu, have_gpu), prefix+".hpp");
+            auto prefix = modfile.filename().replace_extension("").string();
+
+            io::write_all(build_info_header(m, popt, have_cpu, have_gpu), outdir / (mod + ".hpp"));
             for (targetKind target: opt.targets) {
-                std::string outfile = prefix;
                 switch (target) {
-                    case targetKind::gpu:
-                        io::write_all(emit_gpu_cpp_source(m, popt), outfile+"_gpu.cpp");
-                        io::write_all(emit_gpu_cu_source(m, popt), outfile+"_gpu.cu");
+                    case targetKind::gpu: {
+                        fs::path fn = outdir / (mod + "_gpu.cpp");
+                        io::write_all(emit_gpu_cpp_source(m, popt), fn.string());
+                        fn.replace_extension(".cu");
+                        io::write_all(emit_gpu_cu_source(m, popt), fn.string());
                         break;
-                    case targetKind::cpu:
-                        io::write_all(emit_cpp_source(m, popt), outfile+"_cpu.cpp");
+                    }
+                    case targetKind::cpu: {
+                        fs::path fn = outdir / (mod + "_cpu");
+                        fn.replace_extension(".cpp");
```

Comment:
```suggestion
                        fs::path fn = outdir / (mod + "_cpu.cpp");
```

Reply:
I kind of like the symmetry. Yes, it's redundant, but I like it anyway ;)